Add [Targets.Produced]

snowleopard opened the pull request:
A refactoring that makes the code working with directory targets more uniform.

This also fixes a few `Set` modules by exposing a missing type equality.

snowleopard commented on src/dune_engine/build_system.ml:
```diff
@@ -1534,17 +1541,25 @@ end = struct
+                  Path.Build.equal stamp_file
+              in
+              Sandbox.move_targets_to_build_dir sandbox ~loc ~should_be_skipped
+                ~targets
```
Note for reviewers: the change above is slightly non-trivial. I changed the code to pass `should_be_skipped` instead of filtering out files because this works better with the upcoming PR on promoting directory targets.

jeremiedimino commented on src/dune_engine/targets.mli:
```diff
@@ -63,3 +63,37 @@ val head : t -> Path.Build.t option
+  (** Aggregate all content digests. *)
+  val digest : Digest.t t -> Digest.t
+
+  val with_digests :
```
Any reason this is specialised to digests? It seems that we could replace `Digest.t` by `'a` and it would work just as well. It feels that this is a special form of `map`. Maybe it should be called `try_map`?

jeremiedimino commented on src/dune_engine/targets.mli:
```diff
@@ -63,3 +63,37 @@ val head : t -> Path.Build.t option
+  (** Aggregate all content digests. *)
+  val digest : Digest.t t -> Digest.t
+
+  val with_digests :
```
Any reason this is specialised to digests? It seems that we could replace `Digest.t` by `'a` and it would work just as well. It feels that this is a special form of `map`. Maybe it should be called `try_map`?

snowleopard replied:
Yeah, I was hesitating whether to generalise this or not. Probably better to generalise indeed.

snowleopard replied:
Actually, I decided to keep this function non-generalised for now. If we start generalising, the function will need to take one extra parameter, like this:
```ocaml
  val mapi : 'a t -> f:(Path.Build.t -> 'a -> 'b option) -> 'b t option
```
For the current call-sites it means we'll need to deal with an extra `()` which seems like an unnecessary annoyance.

snowleopard replied:
I clarified this in a comment above the function.

snowleopard replied:
OK, after some more thinking I switched to more polymorphic `Option.mapi`.

jeremiedimino commented on src/dune_engine/targets.mli:
```diff
@@ -63,3 +63,37 @@ val head : t -> Path.Build.t option
+module Produced : sig
+  type 'a t = private
+    { files : 'a Path.Build.Map.t
+    ; dirs : 'a Path.Build.Map.t Path.Build.Map.t
```
`'a Path.Build.Map.t Path.Build.Map.t` feels a bit odd. I'd expect `'a String.Map.t Path.Build.Map.t`. Though, maybe it works better this way in the code.

snowleopard replied:
Let me try changing this to `'a String.Map.t Path.Build.Map.t`. I agree that this is better but it was a bit awkward in the first pass. Maybe now it's OK.

snowleopard replied:
OK, I made this change and I think the code got better. Thanks!

jeremiedimino reviewed (changes requested):
Looks good, except for a `raise` that should be `raise_notrace`

snowleopard commented on src/dune_engine/targets.ml:
```diff
@@ -91,3 +91,96 @@ let validate t =
+    let f path () =
+      match f path with
+      | Some digest -> digest
+      | None -> raise Short_circuit
```
Cool, thanks! I didn't know `raise_notrace` existed.

jeremiedimino commented on src/dune_engine/targets.ml:
```diff
@@ -91,3 +91,96 @@ let validate t =
+
+  (* This can be generalised to a more polymorphic [f] if needed. *)
+  let with_digests { files; dirs } ~(f : Path.Build.t -> Digest.t option) =
+    let exception Short_circuit in
```
`let exception X in` is actually not free: it increments a global counter and allocate an extension constructor. It seems better to lift this at the toplevel

snowleopard replied:
Indeed, makes sense!

snowleopard commented:
It looks like something got broken in Dune's incremental mode :( It now hangs for me when building this PR, like this:

```
********** NEW BUILD (src/dune_engine/targets.mli changed) **********

Updating dir_contents cache for "src/dune_engine/targets.mli": Skipped
Updating path_digest cache for "src/dune_engine/targets.mli": Updated { changed = true }
Updating path_stat cache for "src/dune_engine/targets.mli": Skipped
Updating path_exists cache for "src/dune_engine/targets.mli": Skipped
Updating dir_contents cache for "src/dune_engine/targets.mli": Skipped
Updating path_digest cache for "src/dune_engine/targets.mli": Updated { changed = false }
Updating path_stat cache for "src/dune_engine/targets.mli": Skipped
Updating path_exists cache for "src/dune_engine/targets.mli": Skipped
Source files changed, restarting current build... (0/0)
```

@jeremiedimino I wonder if this could be related to the recent cancellation changes?

snowleopard commented:
OK, I addressed all feedback, so this should be ready to merge.

jeremiedimino commented:
> @jeremiedimino I wonder if this could be related to the recent cancellation changes? I think this happens when auto-formatter kicks in a generates a fast file-system event while Dune is building.

On it. I likely broke things with the cancellation PR.

jeremiedimino reviewed (approved):
Looks good!

As discussed offline, i would personally still generalise `with_digest` as the generic version seems easier to understand to me.

jeremiedimino reviewed (approved):
Looks good!
